c++/src/kj/common.h

```diff
@@ -609,6 +610,52 @@ constexpr bool canMemcpy() {
   static_assert(kj::canMemcpy<T>(), "this code expects this type to be memcpy()-able");
 #endif
 
+namespace _ {  // private
+#if defined(__clang__) && defined(__is_identifier)
+#define KJ_HAS_BUILTIN_TYPE_TRAIT(x) (!__is_identifier(x))
+#elif defined(__has_builtin)
+#define KJ_HAS_BUILTIN_TYPE_TRAIT(x) __has_builtin(x)
+#else
+#define KJ_HAS_BUILTIN_TYPE_TRAIT(x) 0
+#endif
+
+#if KJ_HAS_BUILTIN_TYPE_TRAIT(__is_integral)
+template <typename T>
+constexpr bool canMemcmpArrayOf() {
+  return __is_integral(T);
+}
+#else
+template <typename T, typename U>
+constexpr bool isSameType() {
+#if KJ_HAS_BUILTIN_TYPE_TRAIT(__is_same)
+  return __is_same(T, U);
+#elif KJ_HAS_BUILTIN_TYPE_TRAIT(__is_same_as) || __GNUC__
+  return __is_same_as(T, U);
+#else
+  return false;
+#endif
+}
+
+#undef KJ_HAS_BUILTIN_TYPE_TRAIT  // Don't leak this to user code.
+
+template <typename T>
+constexpr bool canMemcmpArrayOfHelper() {
+  return isSameType<T, const bool>()
+      || isSameType<T, const char>()
+      || isSameType<T, const signed char>()   || isSameType<T, const unsigned char>()
+      || isSameType<T, const short int>()     || isSameType<T, const unsigned short int>()
+      || isSameType<T, const int>()           || isSameType<T, const unsigned int>()
+      || isSameType<T, const long int>()      || isSameType<T, const unsigned long int>()
+      || isSameType<T, const long long int>() || isSameType<T, const unsigned long long int>();
+}
+
+template <typename T>
+constexpr bool canMemcmpArrayOf() {
+  return canMemcmpArrayOfHelper<const T>();
```

Comment:
FWIW we have a style pattern of using a `_` suffix to name an internal helper for a template. So I'd rename `canMemcmpArrayOfHelper` to `canMemcmpArray_`.

Although I'm not sure a helper function is needed, you could also do:

```
template <typename T>
constexpr bool canMemcmpArrayOfHelper() {
  typedef const T U;
  return isSameType<U, const bool>()
      || isSameType<U, const char>()
      ...
```

Or alternatively:

```
template <typename T>
constexpr bool canMemcmpArrayOfHelper() {
  typedef RemoveConst<T> U;
  return isSameType<U, bool>()
      || isSameType<U, char>()
      ...
```

Reply:
Oh... there's also an `isIntegral` template defined in `c++/src/kj/units.h`, maybe we should just move it here?

Reply:
I've moved `isIntegral` to `common.h` but it does not support `bool` or `const` types. I don't know if I can change that without affecting other parts of KJ that rely on `isIntegral`, so I'll ignore `bool` and added a workaround for `const` to `operator==`.